photutils/aperture/tests/test_aperture_photometry_3D.py

```diff
@@ -0,0 +1,931 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+The tests in this file test the accuracy of the photometric results.
+Here we test directly with aperture objects since we are checking the
+algorithms in aperture_photometry, not in the wrappers.
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import pytest
+import numpy as np
+from numpy.testing import (assert_allclose, assert_array_equal,
+                           assert_array_less)
+
+from astropy.coordinates import SkyCoord
+from astropy.io import fits
+from astropy.nddata import NDData
+from astropy.table import Table
+from astropy.tests.helper import remote_data
+import astropy.units as u
+from astropy.utils import minversion
+from astropy.wcs.utils import pixel_to_skycoord
+
+from ..core import aperture_photometry
+from ..circle import (CircularAperture, CircularAnnulus, SkyCircularAperture,
+                      SkyCircularAnnulus)
+from ..ellipse import (EllipticalAperture, EllipticalAnnulus,
+                       SkyEllipticalAperture, SkyEllipticalAnnulus)
+from ..rectangle import (RectangularAperture, RectangularAnnulus,
+                         SkyRectangularAperture, SkyRectangularAnnulus)
+from ...datasets import (get_path, make_4gaussians_image, make_wcs,
+                         make_imagehdu, make_4gaussians_cube, make_cubehdu)
+
+try:
+    import matplotlib    # noqa
+    HAS_MATPLOTLIB = True
+except ImportError:
+    HAS_MATPLOTLIB = False
+
+NUMPY_LT_1_14 = not minversion('numpy', '1.14dev')
+
+APERTURE_CL = [CircularAperture,
+               CircularAnnulus,
+               EllipticalAperture,
+               EllipticalAnnulus,
+               RectangularAperture,
+               RectangularAnnulus]
+
+TEST_APERTURES = list(zip(APERTURE_CL, ((3.,), (3., 5.),
+                                        (3., 5., 1.), (3., 5., 4., 1.),
+                                        (5, 8, np.pi / 4),
+                                        (8, 12, 8, np.pi / 8))))
+
+
+# THIS TEST HAS BEEN DONE.
+@pytest.mark.parametrize(('aperture_class', 'params'), TEST_APERTURES)
+def test_outside_array(aperture_class, params):
+    data = np.ones((100, 10, 10), dtype=np.float)
```

Comment:
I wonder whether much simplification (and removing this duplication of tests) can be done by parametrizing the input data array, too, one for the 2D and one for the cube case?